src/smp/ipi.c

```diff
@@ -35,7 +35,8 @@ void ipiStallCoreCallback(bool_t irqPath)
         SCHED_ENQUEUE_CURRENT_TCB;
         switchToIdleThread();
 #ifdef CONFIG_KERNEL_MCS
-        commitTime();
+        /* This is safe because the Idle thread is not schedulable: */
+        chargeBudget(NODE_STATE(ksConsumed), false);
```

Comment:
I can't convince myself that this change is correct either. chargeBudget potentially has to deliver a timeout fault to the thread associated with the ksCurSC if it's run out of budget.  However changing to the IdleThread in the line before means that the ksCurSC isn't the SC of the ksCurThread and so chargeBudget cannot deliver a timeout fault to the correct thread's fault handler.

Reply:
`chargeBudget` is called with `canTimeoutFault == false` here, so it wil never deliver a timeout fault anyway. And as we're handling an IPI I don't think we want to do a timeout fault, as this schedule has nothing to do with timing out.

The whole second if-block will be skipped because the idle thread is not schedulable. Then the only difference with commitTime is a slight detail on how round-robin tasks are charged (ksConsumed versus whole rAmount). If it would go into the second if-block it would trigger `assert(NODE_STATE(ksCurThread)->tcbSchedContext == NODE_STATE(ksCurSC))`.

Reply:
> chargeBudget is called with canTimeoutFault == false here, so it wil never deliver a timeout fault anyway. And as we're handling an IPI I don't think we want to do a timeout fault, as this schedule has nothing to do with timing out.

In the case where the SC is out of budget, the thread it's bound to needs to either be placed in the release queue or generate a timeout fault. Instead, because the thread is added back to the scheduler (on line 35) it could be scheduled again without sufficient budget to be able to leave the kernel and reenter which violates the invariant that all thread's in the scheduler queue must have sufficient budget.

Reply:
It may be worth moving this commit to a separate PR if it's unrelated to moving `updateTimestamp()`?  Is the new test in [this comment](https://github.com/seL4/seL4/pull/847#issuecomment-1136100190) only failing with the `updateTimestamp()` changes, or does it fail for MCS in the main branch too?